orm/src/main/java/org/hibernate/search/query/hibernate/impl/CriteriaObjectInitializer.java

```diff
@@ -105,33 +111,124 @@ private void setCriteriaTimeout(Criteria criteria, TimeoutManager timeoutManager
 		}
 	}
 
-	private void buildUpCriteria(EntityInfo[] entityInfos,
-			Criteria criteria,
-			int maxResults,
-			ObjectInitializationContext objectInitializationContext) {
+	/**
+	 * Returns a list with one or more {@link Criteria} objects for loading the given entity infos. The returned list
+	 * will contain one criteria object for each id space used by the given infos. A single criteria will be returned in
+	 * case all the entity infos originate from the same id space.
+	 */
+	private List<Criteria> buildUpCriteria(EntityInfo[] entityInfos, ObjectInitializationContext objectInitializationContext) {
+		Map<Class<?>, List<EntityInfo>> infosByIdSpace = groupInfosByIdSpace( entityInfos, objectInitializationContext );
+
+		// all entities from same id space -> single criteria
+		if ( infosByIdSpace.size() == 1 ) {
+			// no explicitly user specified criteria query, define one
+			Criteria criteria = objectInitializationContext.getCriteria();
+
+			if ( criteria == null ) {
+				criteria = objectInitializationContext.getSession().createCriteria( infosByIdSpace.keySet().iterator().next() );
+			}
+
+			criteria.add( getIdListCriterion( infosByIdSpace.values().iterator().next(), objectInitializationContext ) );
+
+			return Collections.singletonList( criteria );
+		}
+		// entities originate from different id spaces -> criteria per space
+		else {
+			// Cannot use external criteria for fetching entities from different spaces
+			if ( objectInitializationContext.getCriteria() != null ) {
+				log.givenCriteriaObjectCannotBeApplied();
```

Comment:
Sounds reasonable what you are saying. Not sure whether it is worth the effort. Maybe worth creating a new issue!?